### ISSUE TITLE
Error installing on M1, Ruby 3.x

### ISSUE THREAD

**Issue by btagliani** [kyrylo/rbtree3#16]
Can't install with native extensions, got this error:
```
current directory: /Users/bruno/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/rbtree3-0.7.0
/Users/bruno/.rbenv/versions/3.0.6/bin/ruby -I /Users/bruno/.rbenv/versions/3.0.6/lib/ruby/3.0.0 -r ./siteconf20230407-35327-tin1ri.rb extconf.rb
checking for rb_exec_recursive() in ruby.h... yes
creating Makefile

current directory: /Users/bruno/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/rbtree3-0.7.0
make DESTDIR\= clean

current directory: /Users/bruno/.rbenv/versions/3.0.6/lib/ruby/gems/3.0.0/gems/rbtree3-0.7.0
make DESTDIR\=
compiling dict.c
compiling rbtree.c
rbtree.c:125:14: error: incompatible pointer to integer conversion passing 'const void *' to parameter of type 'VALUE' (aka 'unsigned long') [-Wint-conversion]
    if (TYPE(key1) == T_STRING && TYPE(key2) == T_STRING)
             ^~~~
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/value_type.h:105:52: note: expanded from macro 'TYPE'
#define TYPE(_)           RBIMPL_CAST((int)rb_type(_))
                                                   ^
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/cast.h:33:29: note: expanded from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                            ^~~~
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/value_type.h:178:15: note: passing argument to parameter 'obj' here
rb_type(VALUE obj)
              ^
rbtree.c:125:40: error: incompatible pointer to integer conversion passing 'const void *' to parameter of type 'VALUE' (aka 'unsigned long') [-Wint-conversion]
    if (TYPE(key1) == T_STRING && TYPE(key2) == T_STRING)
                                       ^~~~
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/value_type.h:105:52: note: expanded from macro 'TYPE'
#define TYPE(_)           RBIMPL_CAST((int)rb_type(_))
                                                   ^
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/cast.h:33:29: note: expanded from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                            ^~~~
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/value_type.h:178:15: note: passing argument to parameter 'obj' here
rb_type(VALUE obj)
              ^
rbtree.c:304:18: error: incompatible pointer to integer conversion passing 'const void *' to parameter of type 'VALUE' (aka 'unsigned long') [-Wint-conversion]
        if (TYPE(arg->key) == T_STRING)
                 ^~~~~~~~
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/value_type.h:105:52: note: expanded from macro 'TYPE'
#define TYPE(_)           RBIMPL_CAST((int)rb_type(_))
                                                   ^
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/cast.h:33:29: note: expanded from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                            ^~~~
/Users/bruno/.rbenv/versions/3.0.6/include/ruby-3.0.0/ruby/internal/value_type.h:178:15: note: passing argument to parameter 'obj' here
rb_type(VALUE obj)
              ^
3 errors generated.
make: *** [rbtree.o] Error 1

make failed, exit code 2
```

**Comment by kyrylo** [comment]
Hey Bruno! Unfortunately, I don't have an Apple silicon MacBook. Are you willing to contribute a patch?

**Comment by tsimnujhawj** [comment]
I've opened up a PR on this issue: https://github.com/kyrylo/rbtree3/pull/17

**Comment by PrateekGoyal18** [comment]
@tsimnujhawj I am still facing the same issue as mentioned by @btagliani
Can you help me out in this ?

**Comment by kyrylo** [comment]
Hey @PrateekGoyal18!

Unfortunately, I still don't have an Apple silicon MacBook, and it's hard for me to guess what the problem is.

**Comment by tsimnujhawj** [comment]
> @tsimnujhawj I am still facing the same issue as mentioned by @btagliani
> 
> Can you help me out in this ? 

What Ruby version manager are you using? I switched from `rbenv` to `asdf`, and haven't experienced any gem compiling issues since. You can try that, otherwise, can you share the error log?

**Comment by PrateekGoyal18** [comment]
@tsimnujhawj 
Ruby and rvm versions -
```
ruby -v
ruby 3.2.3 (2024-01-18 revision 52bb2ac0a6) [x86_64-darwin23]

rvm -v
rvm 1.29.12 (latest) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io]
```


Error Logs - 
```
bundle install

Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/rbtree3-0.7.0
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/bin/ruby extconf.rb
checking for rb_exec_recursive() in ruby.h... yes
creating Makefile

current directory: /Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/rbtree3-0.7.0
make DESTDIR\= sitearchdir\=./.gem.20240517-49623-sfpekk sitelibdir\=./.gem.20240517-49623-sfpekk clean

current directory: /Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/rbtree3-0.7.0
make DESTDIR\= sitearchdir\=./.gem.20240517-49623-sfpekk sitelibdir\=./.gem.20240517-49623-sfpekk
compiling dict.c
compiling rbtree.c
rbtree.c:125:14: error: incompatible pointer to integer conversion passing 'const void *' to parameter of
type 'VALUE' (aka 'unsigned long') [-Wint-conversion]
    if (TYPE(key1) == T_STRING && TYPE(key2) == T_STRING)
             ^~~~
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/value_type.h:107:52: note:
expanded from macro 'TYPE'
#define TYPE(_)           RBIMPL_CAST((int)rb_type(_))
                                                   ^
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/cast.h:31:29: note: expanded
from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                            ^~~~
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/value_type.h:224:15: note:
passing argument to parameter 'obj' here
rb_type(VALUE obj)
              ^
rbtree.c:125:40: error: incompatible pointer to integer conversion passing 'const void *' to parameter of
type 'VALUE' (aka 'unsigned long') [-Wint-conversion]
    if (TYPE(key1) == T_STRING && TYPE(key2) == T_STRING)
                                       ^~~~
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/value_type.h:107:52: note:
expanded from macro 'TYPE'
#define TYPE(_)           RBIMPL_CAST((int)rb_type(_))
                                                   ^
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/cast.h:31:29: note: expanded
from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                            ^~~~
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/value_type.h:224:15: note:
passing argument to parameter 'obj' here
rb_type(VALUE obj)
              ^
rbtree.c:304:18: error: incompatible pointer to integer conversion passing 'const void *' to parameter of
type 'VALUE' (aka 'unsigned long') [-Wint-conversion]
        if (TYPE(arg->key) == T_STRING)
                 ^~~~~~~~
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/value_type.h:107:52: note:
expanded from macro 'TYPE'
#define TYPE(_)           RBIMPL_CAST((int)rb_type(_))
                                                   ^
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/cast.h:31:29: note: expanded
from macro 'RBIMPL_CAST'
# define RBIMPL_CAST(expr) (expr)
                            ^~~~
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/value_type.h:224:15: note:
passing argument to parameter 'obj' here
rb_type(VALUE obj)
              ^
rbtree.c:1527:12: warning: 'rb_iterate' is deprecated: by: rb_block_call since 1.9
[-Wdeprecated-declarations]
    return rb_iterate(pp_object_group, (VALUE)&pp_arg,
           ^
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/iterator.h:269:1: note:
'rb_iterate' has been explicitly marked deprecated here
RBIMPL_ATTR_DEPRECATED(("by: rb_block_call since 1.9"))
^
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/include/ruby-3.2.0/ruby/internal/attr/deprecated.h:36:53: note:
expanded from macro 'RBIMPL_ATTR_DEPRECATED'
# define RBIMPL_ATTR_DEPRECATED(msg) __attribute__((__deprecated__ msg))
                                                    ^
1 warning and 3 errors generated.
make: *** [rbtree.o] Error 1

make failed, exit code 2

Gem files will remain installed in /Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/rbtree3-0.7.0 for
inspection.
Results logged to
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/extensions/x86_64-darwin-23/3.2.0/rbtree3-0.7.0/gem_make.out

  /Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:125:in `run'
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:51:in `block
in make'
  /Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:43:in `each'
  /Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:43:in `make'
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/ext_conf_builder.rb:42:in
`build'
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:193:in
`build_extension'
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:227:in
`block in build_extensions'
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:224:in
`each'
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:224:in
`build_extensions'
/Users/prateekgoyal/.rvm/rubies/ruby-3.2.3/lib/ruby/site_ruby/3.2.0/rubygems/installer.rb:852:in
`build_extensions'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/rubygems_gem_installer.rb:71:in
`build_extensions'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/rubygems_gem_installer.rb:28:in
`install'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/source/rubygems.rb:204:in
`install'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/installer/gem_installer.rb:54:in
`install'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/installer/gem_installer.rb:16:in
`install_from_spec'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/installer/parallel_installer.rb:186:in
`do_install'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/installer/parallel_installer.rb:177:in
`block in worker_pool'
  /Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/worker.rb:62:in `apply_func'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/worker.rb:57:in `block in
process_queue'
  /Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/worker.rb:54:in `loop'
  /Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/worker.rb:54:in `process_queue'
/Users/prateekgoyal/.rvm/gems/ruby-3.2.3/gems/bundler-2.3.13/lib/bundler/worker.rb:91:in `block (2
levels) in create_threads'

An error occurred while installing rbtree3 (0.7.0), and Bundler cannot continue.

In Gemfile:
  airbrake was resolved to 10.0.6, which depends on
    airbrake-ruby was resolved to 4.15.0, which depends on
      rbtree3
```

**Comment by kyrylo** [comment]
@PrateekGoyal18 it looks like you are running rbtree3 v0.7.0.

The fix is in v0.7.1. You should update your gem.

**Comment by PrateekGoyal18** [comment]
Hi @kyrylo, yes I just tried to do `gem install rbtree` and it worked - don't know why I didn't tried that before.
Thanks btw!
